src/ess.jl

```diff
@@ -0,0 +1,353 @@
+
+    return BDAESSCache(samples, var)
+end
+
+update_cache!(cache, samples::Matrix, var::Vector) = nothing
+
+function update_cache!(cache::FFTESSCache, samples::Matrix, var::Vector)
+	# check arguments
+    niter, nchains = size(samples)
+    niter == cache.niter || throw(DimensionMismatch())
+    A = cache.A
+    nchains == size(A, 2) || throw(DimensionMismatch())
+	
+    # copy samples and add zero padding
+    n = size(A, 1)
+    T = eltype(A)
+	@inbounds for j in 1:nchains
+		for i in 1:niter
+			A[i, j] = samples[i, j]
+        end
+        for i in (niter + 1):n
+			A[i, j] = zero(T)
+		end
+	end
+
+	# compute unnormalized autocorrelation
+	cache.plan * A
+	@. A = abs2(A)
+	cache.invplan * A
+
+	nothing
+end
+
+# estimation of the autocorrelation function
+function mean_autocorr(k::Int, cache::ESSCache)
+    # check arguments
+    samples = cache.samples
+    niter, nchains = size(samples)
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+	var = cache.var
+    s = zero(eltype(var))
+    firstrange = 1:(niter - k)
+    lastrange = (k + 1):niter
+    @inbounds for i in 1:nchains
+        # increment unnormalized correlation estimates
+        if eltype(samples) isa LinearAlgebra.BlasReal
+            # call into BLAS if possible
+            s += dot(samples, firstrange, samples, lastrange) / var[i]
+            firstrange = firstrange .+ niter
+            lastrange = lastrange .+ niter
+        else
+            # otherwise use views
+            s += dot(view(samples, firstrange, i), view(samples, lastrange, i)) / var[i]
+        end
+    end
+
+    return s / length(samples)
+end
+
+function mean_autocorr(k::Int, cache::FFTESSCache)
+    # check arguments
+    niter = cache.niter
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+    A = cache.A
+    nchains = size(A, 2)
+    s = zero(real(eltype(A)))
+    @inbounds for i in 1:nchains
+        s += real(A[k + 1, i]) / real(A[1, i])
+    end
+
+    return s / nchains
+end
+
+function mean_autocorr(k::Int, cache::BDAESSCache)
+    # check arguments
+    samples = cache.samples
+    niter, nchains = size(samples)
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+	var = cache.var
+    s = zero(eltype(var))
+    @inbounds for j in 1:nchains
+        sj = zero(s)
+        for i in 1:(niter - k)
+            sj += abs2(samples[i, j] - samples[k + i, j])
+        end
+        s += sj / var[j]
+    end
+
+    return 1 - s / (2 * length(samples))
+end
+
+"""
+    ess(chains::Chains; sections = _default_sections(chains), kwargs...)
+
+Estimate the effective sample size and the potential scale reduction.
+"""
+function ess(
+    chains::Chains;
+    sections = _default_sections(chains), # perhaps directly nothing by default - or no sections argument?
```

Comment:
`_default_sections` is mainly for cases where there is no `:parameters` sections in the chain. In the current design of `Chains` (which might change or maybe be extended by an alternate implementation), both the samples and additional information such as log probability and step size are saved in one array but can be distinguished by putting them into different sections. So the logic behind `_default_sections` is that as a default we only want to calculate the ESS for the `:parameters` section if it exists but if there exists no `:parameters` section we want to calculate it for all sections.